README.md
# GPS track record Visualizer (1.1.2)
    Web application that plot GPS track file (.csv) on map and chart.

## Features
1. Center view point of map base on track data.
2. Radius of plots is calculated according to the pace. 
3. Display Pace and Distance on chart.
4. Info of Track (Top pace, distance, time, etc).
5. Highlight plot on Map when clicked on chart, vice-versa.
6. Display info of selected point on Chart/Map.

## Librarys
1. Leaflet.js
2. D3.js
3. Mapbox
4. BootStrap
5. Google Charts

[Demo](http://v2linear.000webhostapp.com/demo/gps-track-virtualization)



changelog.md
# GPS track record Visualizer Changelog

## Version 1.1.2
1. Plot color changes with steps
2. Sample size is configurable 
3. Display info of selected point on Chart/Map

## Version 1.1.1
1. UI improvement
2. Highlight plot on Map when clicked on chart and vice-versa

## Version 1.1.0
1. Track Info section added
2. Head NavBar added
3. UI improvement

## Version 1.0.2
1. Demo mode enabled
2. Distance v.s. Pace Chart Implemented

## Version 1.0.1
1. Read in accessToken from external file
2. Reconstructed file system
3. Able to upload your own csv file

src/functions.js

//check upload file and prepare for d3
function handleFiles(inputFile) {
    inputFile = inputFile[0]; //only accept the 1st one if multi file are uploaded

    var iscsv_file = (function () { //check is uploaded file in csv format
        if (inputFile.name.search(".csv") !== -1) { return true; }
        else {
            print("incorrect file type");
            return false
        };
    })();

    if (iscsv_file) { //continous iff valid csv file is uploaded

        var reader = new FileReader(); //convert csv file into string
        reader.addEventListener("loadend", fileloaded);
        reader.readAsText(inputFile);

        function fileloaded() { //convert raw data into array of object
            csvFileRawData = reader.result;
            csvFileRawData = d3.csvParse(csvFileRawData);
            startAnalysis();
        }

    } else {
        printWarn("invalid csv file");
    }
}

//use demp csv file instead
function demoMode() {
    d3.csv("sample_gps_track_file.csv")
        .get(function (ArrayOfObject) {
            csvFileRawData = ArrayOfObject;
            startAnalysis();
        });
}

//start analysis file and render all the graphes
function startAnalysis() {

    //data processing
    csvFileRawData.forEach(converter); //convert csv data into arrays for d3 operations 
    cal_dis(); //calculate dis between coor and total dis
    cal_pace(csvFileRawData); //calculate pase and store into pase array

    //render everything
    render_chart();
    dis_track_info();
    display_map_chart(); //make map visible
    render_map(); //start rendering map
}

//gather track info & cav raw data 
function dis_track_info() {
    var info_ele = document.getElementById("track_info"),
        cav_dis_ele = document.getElementById("cav_raw_data_textarea"),
        dur_millisecond = (Time_array[Time_array.length - 1] - Time_array[0]);

    info_ele.innerHTML =
        "<b>Duration:</b> " + (dur_millisecond / 60000).toFixed(0) + ":" + (dur_millisecond % 60000 * 60) + "<br>" +
        "<b>Top pace:</b> " + d3.min(pace_array).toFixed(2) + " Min/km " + "<b> Avg pace: </b>" + d3.mean(pace_array).toFixed(2) + " Min/km " + "<br>" +
        "<b>Distance:</b> " + Dis_array[Dis_array.length - 1].toFixed(2) + " km" + "<br>" +
        "<h6>Raw Data:</h6>";
    cav_dis_ele.innerText = d3.csvFormat(csvFileRawData); //convert data array back to string for display
}

//calculate distance between two points (reference: Haversine Formula)
function dis_Lat_Lot(lat1, lat2, lon1, lon2) {
    Number.prototype.toRad = function () {
        return this * Math.PI / 180;
    }

    var R = 6371; // radius of the Earth
    var x1 = lat2 - lat1;
    var dLat = x1.toRad();
    var x2 = lon2 - lon1;
    var dLon = x2.toRad();
    var a = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
        Math.cos(lat1.toRad()) * Math.cos(lat2.toRad()) *
        Math.sin(dLon / 2) * Math.sin(dLon / 2);
    var c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));

    return R * c;


}

src/map_functions.js

//compute pace and store in pace_array
function cal_pace(dataSet) {
    var avg_pace, points_dis = 0, points_dur; //distance and duration between sample_size points

    for (var pos = 0; pos < dataSet.length - pace_sample_size; pos++) {

        if (pos % pace_sample_size == 0 && pos != 0) {
            points_dur = (Time_array[pos].getHours() * 60 + Time_array[pos].getMinutes() + Time_array[pos].getSeconds() / 60) - //duration between 1st and last sample size points
                        (Time_array[pos - pace_sample_size].getHours() * 60 + Time_array[pos - pace_sample_size].getMinutes() + Time_array[pos - pace_sample_size].getSeconds() / 60);
            
            points_dis = Dis_array[pos] - Dis_array[pos - pace_sample_size];

            pace_array.push(
                points_dur / points_dis
            );
        }
    }

    for (var pos2 = dataSet.length - pace_sample_size; pos2 < dataSet.length; pos2++) { //make up the last pace data at the end
        pace_array.push(pace_array[pace_array.length - 1]);
    }

}

// convert Lat & Lot into Distance and cal total distance
function cal_dis() {
    for (var pos = 0; pos < Lot_array.length - 1; pos++) {
        var dis = dis_Lat_Lot(Lot_array[pos], Lot_array[pos + 1], Lat_array[pos], Lat_array[pos + 1]);
        Dis_array.push(total_dis);
        total_dis += dis;
    }
}

//convert raw data into float and Date format & store into individual arrays for d3 operation
function converter(record_ele) {
    Lot_array.push(parseFloat(record_ele.Y));
    Lat_array.push(parseFloat(record_ele.X));

    //convert raw time into date object in javascript
    var raw_string = record_ele.time;

    //extracting time pieces form raw data & push into time array
    Time_array.push(new Date(
        raw_string.slice(0, 4),    //Year
        raw_string.slice(5, 7),    //Month
        raw_string.slice(8, 10),   //Day
        raw_string.slice(11, 13),  //Hour
        raw_string.slice(14, 16),  //Second
        raw_string.slice(17, 19),
        "00"
    ));

    return record_ele;
}

//readin accessToken from external file for mapbox
function readin_accessToken(URL) {
    var raw_data = "";
    $.ajax({
        url: URL,
        success: function (data) {
            raw_data = data;
        },
        async: false
    });

    if (raw_data == "") { //accessToken is missing
        printWarn("AccessToken is missing");
    } else return raw_data;
}

//make map & chart visible 
function display_map_chart() {
    document.getElementById("fileInput").style.visibility = "hidden";//hide buttons
    document.getElementById("demoBtn").style.visibility = "hidden";
    var browser_height = $(window).height(),
        browser_width = $("#navBar").width(),
        topRow_height = document.getElementById("navBar").offsetHeight + 
                        document.getElementById("topRow").offsetHeight,
        map_obj = document.getElementById("map");

    map_obj.style.height = parseFloat(browser_height - topRow_height) * 0.99 + "px"; //set height and width for map object
    map_obj.style.width = parseFloat(browser_width) + "px"; 
    map_obj.style.visibility = "visible";
    document.getElementById("cav_raw_data_textarea").style.visibility = "visible" //make raw data win visible
}

//render map in DOM
function render_map() {

    L.mapbox.accessToken = readin_accessToken("src/accessToken.txt");

    map = L.mapbox.map('map', 'mapbox.streets')
        .setView([d3.mean(Lot_array),
        d3.mean(Lat_array)],
        14); //init map and set center viewpoint accroding to data set

    radius_scale = d3.scaleLinear();
    radius_scale.range([0, 15]);
    radius_scale.domain([d3.min(pace_array), d3.max(pace_array)]);

    //plotting track
    for (var pos = 0; pos < Lot_array.length - 1; pos++) {
        L.circle([Lot_array[pos], Lat_array[pos]], {
            color: 'red',
            fillColor: '#f03',
            fillOpacity: 0.8,
            radius: radius_scale(pace_array[Math.round(pos / pace_sample_size)])
        })
            .addTo(map);
    }
}


src/chart_functions.js
//all operation of rendering chart
function render_chart() {

    var chartDataSet = [], 
        time_label_array = [], 
        numOfLable_horiz = 7; //number of label that render on horiz axis

    //processing data for chart rendering
    pace_sample_size += 1;
    for (var pos = 0; pos < Lot_array.length; pos++) {
        var pace_data = function () { //make up the empty data slot by placing pevious pace data
            if (pos % pace_sample_size != 0) {
                return pace_array[Math.round(pos / pace_sample_size)];
            } else {
                return pace_array[pos];
            }
        }();

        if(pos % Math.round(Time_array.length / numOfLable_horiz) == 0){ //only store hour and minute
            time_label_array.push((Time_array[pos].getHours().toString() + ":" + Time_array[pos].getMinutes().toString()));
        }
    
        var push_data = [ //construct feed in data set
            (Time_array[pos].getHours().toString() + ":" + Time_array[pos].getMinutes().toString() + ":" + Time_array[pos].getSeconds().toString()),
            Dis_array[pos],
            pace_data
        ];

        chartDataSet.push(push_data);
    }

    google.charts.setOnLoadCallback(function () {

        var data = new google.visualization.DataTable();

        data.addColumn('string', 'Time');
        data.addColumn('number', 'Distance');
        data.addColumn('number', 'Pace');
        data.addRows(chartDataSet);

        var options = { //options for chart
            series: {
                0: { targetAxisIndex: 0 },
                1: { targetAxisIndex: 1 }
            },
            vAxes: {
                0: { title: 'Distance (KM)' },
                1: { title: 'Pace (Min/KM)' }
            },
            hAxis: {
                showTextEvery : Math.round(Time_array.length / numOfLable_horiz),
                fontSize: 6
            }
        };

        var chart = new google.visualization.LineChart(document.getElementById('chartRow'));  // Instantiate and draw the chart.
        chart.draw(data, options);
    });
}


src/data_fileOperations_functions.js

//compute pace and store in pace_array
function cal_pace(dataSet) {
    var avg_pace, points_dis = 0, points_dur; //distance and duration between sample_size points

    for (var pos = 0; pos < dataSet.length - pace_sample_size; pos++) {
        if (pos % pace_sample_size == 0 && pos != 0) {
            points_dur = (Time_array[pos].getHours() * 60 + Time_array[pos].getMinutes() + Time_array[pos].getSeconds() / 60) - //duration between 1st and last sample size points
                (Time_array[pos - pace_sample_size].getHours() * 60 + Time_array[pos - pace_sample_size].getMinutes() + Time_array[pos - pace_sample_size].getSeconds() / 60);

            points_dis = Dis_array[pos] - Dis_array[pos - pace_sample_size];

            pace_array.push(
                points_dur / points_dis
            );
        }else if(pos != 0){
            pace_array.push(
                pace_array[pace_array.length - 1]
            );
        }
    }

    //make up the last "pace_sample_size" ele
    for(var pos = dataSet.length - pace_sample_size; pos < dataSet.length; pos++){
      pace_array.push(pace_array[pace_array.length - 1]);

    }

    //place back missing ele at the beginning, since the first "pace_sample_size" elements can't be calculated until pointer reach "pace_sample_size"
    for(var pos = pace_sample_size - 1; pos >= 0; pos--){
        pace_array[pos] = pace_array[pace_sample_size];
    }
}

// convert Lat & Long into Distance and cal total distance
function cal_dis() {
    for (var pos = 0; pos < Long_array.length - 1; pos++) {
        var dis = dis_Lat_Long(Long_array[pos], Long_array[pos + 1], Lat_array[pos], Lat_array[pos + 1]);
        Dis_array.push(total_dis);
        total_dis += dis;
    }
}

//convert raw data into float and Date format & store into individual arrays for d3 operation
function converter(record_ele) {
    Long_array.push(parseFloat(record_ele.Y));
    Lat_array.push(parseFloat(record_ele.X));

    //convert raw time into date object in javascript
    var raw_string = record_ele.time;

    //extracting time pieces form raw data & push into time array
    Time_array.push(new Date(
        raw_string.slice(0, 4),    //Year
        raw_string.slice(5, 7),    //Month
        raw_string.slice(8, 10),   //Day
        raw_string.slice(11, 13),  //Hour
        raw_string.slice(14, 16),  //Second
        raw_string.slice(17, 19),
        "00"
    ));

    return record_ele;
}

//readin accessToken from external file for mapbox
function readin_accessToken(URL) {
    var raw_data = "";
    $.ajax({
        url: URL,
        success: function (data) {
            raw_data = data;
        },
        async: false
    });

    if (raw_data == "") { //accessToken is missing
        printWarn("AccessToken is missing");
    } else return raw_data;
}

//start analysis file and render all the graphes
function startAnalysis() {

    //data processing
    csvFileRawData.forEach(converter); //convert csv data into arrays for d3 operations 
    cal_dis(); //calculate dis between coor and total dis
    cal_pace(csvFileRawData); //calculate pase and store into pase array

    //render everything
    render_chart();
    dis_track_info();
    display_map_chart(); //make map visible
    render_map(); //start rendering map
}

//calculate distance between two points (reference: Haversine Formula)
function dis_Lat_Long(lat1, lat2, lon1, lon2) {
    Number.prototype.toRad = function () {
        return this * Math.PI / 180;
    }

    var R = 6371, // radius of the Earth
        x1 = lat2 - lat1,
        dLat = x1.toRad(),
        x2 = lon2 - lon1,
        dLon = x2.toRad(),
        a = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
            Math.cos(lat1.toRad()) * Math.cos(lat2.toRad()) *
            Math.sin(dLon / 2) * Math.sin(dLon / 2),
        c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));

    return R * c;
}

//check upload file and prepare for d3
function handleFiles(inputFile) {
    inputFile = inputFile[0]; //only accept the 1st one if multi file are uploaded

    var iscsv_file = (function () { //check is uploaded file in csv format
        if (inputFile.name.search(".csv") !== -1) { return true; }
        else {
            print("incorrect file type");
            return false
        };
    })();

    if (iscsv_file) { //continous iff valid csv file is uploaded

        var reader = new FileReader(); //convert csv file into string
        reader.addEventListener("loadend", fileloaded);
        reader.readAsText(inputFile);

        function fileloaded() { //convert raw data into array of object
            csvFileRawData = reader.result;
            csvFileRawData = d3.csvParse(csvFileRawData);
            startAnalysis();
        }

    } else {
        printWarn("invalid csv file");
    }
}

//use demp csv file instead
function demoMode() {
    d3.csv("sample_gps_track_file.csv")
        .get(function (ArrayOfObject) {
            csvFileRawData = ArrayOfObject;
            startAnalysis();
        });
}

src/ui_functions.js
//make map & chart visible 
function display_map_chart() {

    //hide nav bar buttons
    document.getElementById("fileInput").style.visibility = "hidden";
    document.getElementById("demoBtn").style.visibility = "hidden";
    document.getElementById("Info_panel_body").style.height = "100%"; //reset height after data is loaded

    var mapPanel_width = $("#map_panel").width(),
        map_obj = document.getElementById("map");

    map_obj.style.width = parseFloat(mapPanel_width) + "px";
    map_obj.style.height = parseFloat(mapPanel_width * 0.5) + "px"; //set height and width for map object

    //set the height of the right panel the same as the left one
    document.getElementById("Info_panel_body").style.height =
        $("#Map_panel_body").outerHeight(true) + "px";


    map_obj.style.visibility = "visible";
    document.getElementById("tabs").style.visibility = "visible";
    document.getElementById("cav_raw_data_textarea").style.visibility = "visible" //make raw data win visible
}

//gather track info & cav raw data 
function dis_track_info() {
    var info_ele = document.getElementById("track_info"),
        cav_dis_ele = document.getElementById("cav_raw_data_textarea"),
        dur_millisecond = (Time_array[Time_array.length - 1] - Time_array[0]);

    info_ele.innerHTML =
        "<b>Duration:</b> " + (dur_millisecond / 60000).toFixed(0) + ":" + (dur_millisecond % 60000 * 60) + "<br>" +
        "<b>Top pace:</b> " + d3.min(pace_array).toFixed(2) + " Min/km " + "<b> Avg pace: </b>" + d3.mean(pace_array).toFixed(2) + " Min/km " + "<br>" +
        "<b>Distance:</b> " + Dis_array[Dis_array.length - 1].toFixed(2) + " km" + "<br>" +
        "<h6>Raw Data:</h6>";
    cav_dis_ele.innerText = d3.csvFormat(csvFileRawData); //convert data array back to string for display
}

//render map in DOM
function render_map() {

    L.mapbox.accessToken = readin_accessToken("src/accessToken.txt");

    map = L.mapbox.map('map', 'mapbox.streets')
        .setView([d3.mean(Long_array),
        d3.mean(Lat_array)],
        14); //init map and set center viewpoint accroding to data set

    var radius_scale = d3.scaleLinear(), color_scale = d3.scaleLinear();
    radius_scale.range([0, 20]);
    radius_scale.domain([d3.min(pace_array), d3.max(pace_array)]);
    color_scale.range(["yellow", "green"]);
    color_scale.domain([0, Long_array.length]);

    //plotting track
    for (var pos = 0; pos < Long_array.length - 1; pos++) {
        var render_circle =
            L.circle([Long_array[pos], Lat_array[pos]], {
                color: color_scale(pos), //progressive color changes
                fillColor: color_scale(pos),
                fillOpacity: 0.8,
                radius: radius_scale(pace_array[pos]),
                pos_in_data: pos //store pos in data in object
            });

        render_circle.addTo(map); //add plot to map
        render_circle.on('click', function (e) { //handler of click events
            clicked_on_map(e.target.options.pos_in_data);
        })
    }
}

//Highlight point on chart and map
function highlight_point(pos_in_data) {

    if (last_plot_obj != null) last_plot_obj.remove(); //remove marker on map

    var loc_x = chart_interactive_interface.getXLocation(pos_in_data),
        marker = document.getElementById("overlay-marker"),
        chart_ele_style = document.getElementsByTagName("rect")[0].getBoundingClientRect(),
        loc_x_marker = Math.floor(loc_x);

    //moving marker to target position
    marker.style.visibility = "visible";
    marker.style.height = chart_ele_style.bottom - chart_ele_style.top + "px";
    marker.style.left = loc_x_marker + 30 + "px";
    marker.style.top = chart_ele_style.top + "px";

    //add marker on map as well
    if (last_plot_obj != null) last_plot_obj.remove(); //remove last marker
    last_plot_obj = L.marker([Long_array[pos_in_data], Lat_array[pos_in_data]]).addTo(map);

    //display selected data 
    display_sel_data(pos_in_data);

}

//display selected data on "sel data" panel
function display_sel_data(pos_in_data){
    var display_ele = document.getElementById("sel_data_info");
    display_ele.innerHTML = 
        "<b>Distance: </b>" + Dis_array[pos_in_data].toFixed(2) + " km<br>" +
        "<b>Pace: </b>" + pace_array[pos_in_data].toFixed(2) + " Min/km <br>" +
        "<b>Time: </b>" + Time_array[pos_in_data].getHours() + ":" + Time_array[pos_in_data].getMinutes() + ":" + Time_array[pos_in_data].getSeconds() + "<br>" +
        "<b>Lat: </b>" + Lat_array[pos_in_data] + "<b> Long: </b>" + Long_array[pos_in_data];

    $('#tabs a:last').tab('show')
}

//Highlight point on chart 
function clicked_on_map(pos_in_data){
    chart_obj.setSelection([]);
    highlight_point(pos_in_data);
}

//update position of highlight bar on chart if scrolled (to prevent mispositioning)
function update_highlight_bar_pos() {
    var marker = document.getElementById("overlay-marker"),
        chart_ele_style = document.getElementsByTagName("rect")[0].getBoundingClientRect();

    marker.style.top = chart_ele_style.top + "px";
}

//all operation of rendering chart
function render_chart() {

    var chartDataSet = [],
        time_label_array = [],
        numOfLable_horiz = 7; //number of label that render on horiz axis

    //processing data for chart rendering
    pace_sample_size += 1;
    for (var pos = 0; pos < Long_array.length; pos++) {

        var pace_data = pace_array[pos];

        if (pos % Math.round(Time_array.length / numOfLable_horiz) == 0) { //only store hour and minute
            time_label_array.push((Time_array[pos].getHours().toString() + ":" + Time_array[pos].getMinutes().toString()));
        }

        var push_data = [ //construct feed in data set
            (Time_array[pos].getHours().toString() + ":" + Time_array[pos].getMinutes().toString() + ":" + Time_array[pos].getSeconds().toString()),
            Dis_array[pos],
            pace_data
        ];

        chartDataSet.push(push_data);
    }

    pace_sample_size--; //reset sample size after calculation

    //start loadin chart
    google.charts.setOnLoadCallback(function () {
        var data = new google.visualization.DataTable();

        data.addColumn('string', 'Time');
        data.addColumn('number', 'Distance');
        data.addColumn('number', 'Pace');
        data.addRows(chartDataSet);

        var options = { //options for chart
            series: {
                0: { targetAxisIndex: 0 },
                1: { targetAxisIndex: 1 }
            },
            vAxes: {
                0: { title: 'Distance (KM)' },
                1: { title: 'Pace (Min/KM)' }
            },
            hAxis: {
                showTextEvery: Math.round(Time_array.length / numOfLable_horiz),
                fontSize: 6
            }
        };

        chart_obj = new google.visualization.LineChart(document.getElementById('chartRow'));  // Instantiate and draw the chart.

        //add handler & listener for "select" event
        function selectHandler() {
            var selectedItem = chart_obj.getSelection()[0];
            if (selectedItem) {
                highlight_point(selectedItem.row); //pass selected data point to function that plot on map
            }
        }

        //add handler & listener after chart is rendered, and store obj of interactive_interface_chart for later operation 
        function get_interactive_interface_chart(dataTable) {
            chart_interactive_interface = this.getChartLayoutInterface();
        };

        //add listener
        google.visualization.events.addListener(chart_obj, 'select', selectHandler);
        google.visualization.events.addListener(chart_obj, 'ready', get_interactive_interface_chart.bind(chart_obj, data));

        //render chart
        chart_obj.draw(data, options);
    });
}

//function handle sample size editing
function sample_size_change(mode) {
    function update_sample_size_input() {
        document.getElementById("sample_size_input").value = pace_sample_size;
    }
    switch (mode) {
        case "add":
            pace_sample_size++;
            update_sample_size_input();
            break;

        case "sub":
            if (pace_sample_size > 1) {
                pace_sample_size--;
                update_sample_size_input();
            }
            break;
    }
}

//function handle redering and recalculation after variable changes
function re_process() {

    //remove rendered DOM
    map.remove();

    //reset all data sets
    Long_array = []; Lat_array = []; Time_array = []; pace_array = []; Dis_array = []; total_dis = 0;

    //restart calculation and rendering process
    startAnalysis();
}
